### ISSUE TITLE
External backup of NodeBB

### ISSUE THREAD

**Issue by stevefan1999-personal** [NodeBB/NodeBB#12215]
### Description

We should be able to get a simple backup/restore procedure externally, by dumping the database into SQL files and saving the config.json and package.json into a gzip file. This will greatly improve the lifecycle experience for operating a NodeBB instance in a safe and secure way.

Part of #12213, also for #12178

**Comment by julianlam** [comment]
Hi @stevefan1999-personal, it's been awhile since we approached this issue actually. So long that when we first talked about it, Redis was the only database we supported.

Back then we declined to support this for a simple reason: Redis does not have a command accessible to NodeBB to export the database. You can `SAVE` and `BGSAVE`, but the actual database files are inaccessible to a properly configured (aka unprivileged) NodeBB instance.

**Comment by oplik0** [comment]
For redis, you can scan+dump and then restore. It needs some more manual work but should be fine.

Postgres has `pg_dump` (and `pg_dumpall` which exports all databasees) similar to `mongodump`.

However, I'd argue that using DB-specific backup methods for software that supports multiple databases has a fundamental flaw - you can only restore on similar setups. No cross-db migration using backups. While probably slower (and requiring a bit more work), considering that NodeBB already has essentially its own redis-inspired database abstraction layer, using this abstraction for a custom backup solution might be a better idea.

**Comment by julianlam** [comment]
The other argument was that it shouldn't be the responsibility of the forum software to handle backup and restore as there is more than adequate tooling to do this at the system level.

Anything we'd do would merely be an abstraction over existing tooling (with exception of @oplik0's suggestion of having a bespoke export format)

**Comment by oplik0** [comment]
Yup. Just writing a script that calls `mongodump` and `rsync`'s it (or sends in any other way) somewhere along with `config.json` and `package.json` is simple enough that I think most people looking into rolling a custom backup solution (instead of relying on some full disk backups from a cloud provider or something) should probably be able to write one. And it really is a lot less work to make something setup-specific than writing a general solution.

So I agree that a wrapper over few db tools and gzip is kind of out of scope, but a bespoke export might be interesting and it could further help decouple NodeBB from the underlying database further (since it'd allow for backup+restore for migrations between DBs)

**Comment by mySebbe** [comment]
A generic `nodebb backup` wrapper still cannot provide one reliable contract across MongoDB, PostgreSQL, and Redis. Native `mongodump`/`pg_dump` and platform snapshots already define consistency and restore behavior; Redis and uploaded files require different handling. I suggest closing the original wrapper proposal. If cross-database migration is the real goal, that should be a new export/import format issue specifying quiescing, schema/version metadata, uploads, integrity verification, and a tested restore path rather than calling it a backup.
